meteograph.1
.TH METEOGRAPH "1" "December 2001" "Meteo station tools" Othello
.SH NAME
meteograph \- draw graphs of meteo data
.SH SYNOPSIS
meteograph [
.B \-a
] [
.B \-c
] [
.B \-d
] [
.BI \-e endtime
]
.BI \-f conffile
[
.BI \-g graphname
|
.BI \-G graphname
]
.IR interval ...
.SH DESCRIPTION
The 
.BR meteograph (1)
program computes graphs of meteo data for a weather station for various
sampling
.IR interval s.
Allowed values for the
.I interval
are 300, 1800, 7200 and 86400, resulting in graphs spanning a day,
a week, a month and a full year, respectively.
By default, the graph ends at the current point in time, but this
behaviour can be modified by specifying an end timestamp with
the 
.B \-e
option in the format
.IR YYYYMMDDhhmmss .

By default, all graphs are produced, i.e. for temperature, inside
temperature, barometric pressure, wind, rain and radiation.
The
.B \-g
and
.B \-G
options allow to modify the list of graphs to be produced.
The graphs are stored in files in the current working directory
unless a different directory is specified with the 
.B \-c
option. The file names are of the
form

.ce
.IR station - graphname - suffix .png

The
.I station
name is specified in the configuration file.
The suffix is one of
.IR day ,
.IR week ,
.I month
or
.IR year ,
for sampling intervals 300, 1800, 7200 and 86400 respectively.
The possible  graph names are summarized in the following  table

.TS
tab(&);
l l.
temperature&outside temperature
temperature_inside&inside temperature
barometer&barometric pressure
wind&wind speed and direction
rain&rain total
radiation&solar and uv radiation
.TE

The parameters defining these graphs are set in the
.BR meteo.conf (5)
file.

.SH OPTIONS
.TP
.B \-a
use the 
.I averages
table to compute the graphs. This requires that 
.BR meteoavg (1)
is run either as a daemon or before
.BR meteograph (1)
to ensure all necessary averages are present in the
.I averages
table.
.TP
.BI \-c imagedir
write image files to 
.IR imagedir .
Note that 
.BR meteograph (1)
performs a
.BR chdir (2)
before opening the configuration file (as specified with the
.B \-f 
option below), so if a relative path for the configuration file
is used, it should be relativ to the image directory.
One possible application of this feature is to have a graph directory
for each weather station, and to put the meteo.conf
file in the same directory as the graphs. 
.BR meteograph (1) can then be called with

.nf
.ti +3
meteograph -c graphdir-a -f meteo.conf 7200
.ti +3
meteograph -c graphdir-b -f meteo.conf 7200
.fi

to reproduce all two hour graphs for stations
.I a
and
.IR b .
.TP
.B \-d
increase the debug level by one. Debug messages are written to standard
error.
.TP
.BI \-e endtime
compute graphs ending at
.IR endtime .
.TP
.BI \-f conffile
Use 
.I conffile
to read the configuration, see 
.BR meteo.conf (5)
for the configuration syntax. This argument is mandatory.
.TP
.BI \-g graphname
add a graph named
.I graphname
to the list of graphs to be produced. If this option is specified
once, no other graphs are produced than the ones specified.
.TP
.BI \-G graphname
suppress the graph named
.IR graphname ,
but otherwise produce all graphs.
If this option is used several times, all graphs specified are
suppressed.

.SH "SEE ALSO"
.BR meteo (1)
.BR meteo.conf (5),
.BR meteoloop (1),
.BR meteoavg (1),
.BR meteodequeue (1),
.BR meteocheck (1),
.BR meteojob (1)

.SH AUTHOR
Dr. Andreas Mueller <afm@othello.ch>


NEWS
0.9.17	a new outlet interface can be used to process meteo station data
	directly. In contrast to the map interface, the data is pushed
	to a file or into a pipe by the meteopoll process every minute.

0.9.15	new interface which allows to read current readings from the console
	to display on the web page. The new -m option to meteopoll makes
	sure current data is written to a file (the argument to the option)
	in XML format. Since updates to this file are very frequent, one
	should read the file atomically, one should use the meteomap program
	to read the data.
	fix of a nasty bug in the week label code, causing navigation failure
	at week resolution in some cases

0.9.14	gallery files included as samples
	can now compute sunrise and sunset (through meteosunriset and Sun.cc)

0.9.13	minor bug in meteobrowser.php
	gallery pages

0.9.12	some minor fixes, plus a security fix: remote code execution in
	meteobrowser.php because of missing escapeshellcmd

0.9.11	some bug fixes for GroWeather

0.9.10	turn of CRC0 for GroWeather and possibly other type E weatherlinks.

0.9.9	A hack to protect from IE reading the complete meteobrowser.php
	space offline

0.9.8	fixes a integer overflow bug: to compute the time labels for the
	maps, GraphWindow.cc used the formular (start + end)/2, but since
	about the middle of january, this causes an overflow, so that the
	target ends up in february 1936. Replace that by start/2 + end/2.
	
0.9.7	some bug fixes
	removal of offset configuration variable
	added averages configuration in XML file, so the user now gets full
	control about which averages are compute, and how
	battery and transmitter fields now work
	sunrise and sunset can also be stored in the database (this is a
		waste of space, but who cares about a few gibabytes these
		days)
	added channel of type sun, which indicates the time during which
		the sun is above the horizon

0.9.6	some bug fixes
	removal of header table

0.9.5	some minor fixes

(for further news please check the ChangeLog)

0.9.0	Major rewrite of many aspects of meteo sponsored by the University
	Univates in Lajeado in Rio Grande do Sul, Southern Brazil. Further
	contributions: Jeff Wabik (new derived value computations).

--
$Id: NEWS,v 1.32 2009/01/10 19:00:23 afm Exp $


src/meteograph.cc
/*
 * meteograph.cc -- create meteo data graphs similar to MRTGs
 *
 * (c) 2001 Dr. Andreas Mueller, Beratung und Entwicklung
 *
 * $Id: meteograph.cc,v 1.4 2003/06/12 23:29:46 afm Exp $
 */
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#include <meteo.h>
#include <meteograph.h>
#include <xmlconf.h>
#include <graph.h>
#include <mysql.h>
#ifdef HAVE_STDIO_H
#include <stdio.h>
#endif
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#include <math.h>
#include <dewpoint.h>
#ifdef HAVE_STRING_H
#include <string.h>
#endif
#include <errno.h>
#include <time.h>
#include <database.h>
#include <dograph.h>
#include <printver.h>
#include <mdebug.h>
#include <dewpoint.h>

extern int	optind;
extern char	*optarg;

typedef struct namegraph_s {
	char	*name;
	int	graph;
} namegraph_t;

#define	NGRAPHS	14
static namegraph_t	ngs[NGRAPHS] = {
	{ "barometer",		DOGRAPH_BAROMETER			},
	{ "pressure",		DOGRAPH_BAROMETER			},
	{ "temperature",	DOGRAPH_TEMPERATURE			},
	{ "thermo",		DOGRAPH_TEMPERATURE			},
	{ "temperature_inside",	DOGRAPH_TEMPERATURE_INSIDE		},
	{ "thermo_inside",	DOGRAPH_TEMPERATURE_INSIDE		},
	{ "humidity",		DOGRAPH_HUMIDITY			},
	{ "hygro",		DOGRAPH_HUMIDITY			},
	{ "humidity_inside",	DOGRAPH_HUMIDITY_INSIDE			},
	{ "hygro_inside",	DOGRAPH_HUMIDITY_INSIDE			},
	{ "rain",		DOGRAPH_RAIN				},
	{ "precipitation",	DOGRAPH_RAIN				},
	{ "wind",		DOGRAPH_WIND				},
	{ "radiation",		DOGRAPH_RADIATION			}
};

/* declare to keep compiler happy					*/
void	usage(void);

void	usage(void) {
	printf(
"usage: meteograph [ -adtV ] [ -l logfile ] [ -c dir ] [ -G graph ] \n"
"[ -g graph ] [ -f config ] [ -e endtime ] [ -V ]\n"
"	-a		use averages table\n"
"	-d		increase debug level\n"
"	-t		include timestamp in generated file name\n"
"	-s stationname	name of the station for which to compute graphs\n"
"	-V		display version and exit\n"
"	-l logfile	write debug log to logfile\n"
"	-c dir		change directory to dir before writing any image files\n"
"	-g graph	include only graph (can be given multiple times)\n"
"	-G graph	exclude graph (can given multiple times)\n"
"			[ -g and -G can not be used simulatineouly ]\n"
"	-f config	use configuration file config\n"
"	-e endtime	specify end time of graph\n");
}

int	main(int argc, char *argv[]) {
	char		*conffilename = METEOCONFFILE;
	char		*station = NULL;
	int		c, interval, i, G = 0;
	dograph_t	dg;
	time_t		t;
	struct tm	*tp;
	char		timestamp[64];
	meteoconf_t	*mc;

	/* initialize defaults						*/
	time(&dg.end);
	dg.useaverages = 0;
	dg.requestedgraphs = 0;
	dg.withtimestamps = 0;
	dg.timestamp = NULL;
	
	/* parse command line						*/
	while (EOF != (c = getopt(argc, argv, "adc:e:f:g:l:G:ts:V?")))
		switch (c) {
		case '?':
			usage();
			exit(EXIT_SUCCESS);
			break;
		case 'l':
			if (mdebug_setup("meteograph", optarg) < 0) {
				fprintf(stderr, "%s: cannot init log\n",
					argv[0]);
				exit(EXIT_FAILURE);
			}
			break;
		case 'c':
			if (chdir(optarg) < 0) {
				mdebug(LOG_CRIT, MDEBUG_LOG, 0,
					"cannot chdir to %s", optarg);
				exit(EXIT_FAILURE);
			}
			break;
		case 'd':
			debug++;
			break;
		case 'e':
			/* convert a timestamp into a time_t		*/
			dg.timestamp = optarg;
			dg.end = localtime2time(optarg);
			if (debug)
				printf("%s:%d: timestamp %d\n", __FILE__,
					__LINE__, (int)dg.end);
			break;
		case 't':
			/* include time stamp in generated file		*/
			dg.withtimestamps = 1;
			break;
		case 'f':
			conffilename = optarg;
			break;
		case 'a':
			dg.useaverages = 1;
			if (debug)
				mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
					"use of average table requested");
			break;
		case 'g':
			if (debug)
				mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
					"graph for %s requested", optarg);
			for (i = 0; i < NGRAPHS; i++) 
				if (0 == strcasecmp(optarg, ngs[i].name))
					dg.requestedgraphs |= ngs[i].graph;
			break;
		case 'G':
			if (!G) { dg.requestedgraphs = 0xff; G = 1; }
			for (i = 0; i < NGRAPHS; i++) 
				if (0 == strcasecmp(optarg, ngs[i].name))
					dg.requestedgraphs ^= ngs[i].graph;
			break;
		case 's':
			station = optarg;
			break;
		case 'V':
			fprintver(stdout);
			exit(EXIT_SUCCESS);
			break;
		default:
			fprintf(stderr, "%s:%d: unknown option: %c\n",
				__FILE__, __LINE__, c);
			usage();
			exit(EXIT_FAILURE);
		}

	/* check for requested graphs					*/
	if (dg.requestedgraphs == 0) {
		if (debug)
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "no graphs selected, "
				"creating all");
		dg.requestedgraphs = 0xff;
	}
	if (debug)
		mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "graph mask: %02x",
			dg.requestedgraphs);

	/* station name is mandatory					*/
	if (NULL == station) {
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "must specify -s <station>");
		exit(EXIT_FAILURE);
	}

	/* it is an error not to specify a configuration file		*/
	if (NULL == conffilename) {
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "must specify -f <config>");
		exit(EXIT_FAILURE);
	}

	/* read the configuration file					*/
	if (NULL == (mc = xmlconf_new(conffilename, station))) {
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "configuration invalid");
		exit(EXIT_FAILURE);
	}

	/* find the temperature unit to make sure dew point is computed	*/
	/* correctly							*/
	find_tempunit(mc, station);

	/* open the database connection					*/
	dg.mysql = mc_opendb(mc, O_RDONLY);
	if (dg.mysql == NULL) {
		mdebug(LOG_CRIT, MDEBUG_LOG, 0,
			"could not connect to database");
		exit(EXIT_FAILURE);
	}

	/* get some other data from the configuration file		*/
	dg.prefix = mc->station;

	/* if the timestamp is not set, we take the current time for 	*/
	/* timestamp							*/
	if (NULL == dg.timestamp) {
		time(&t);
		tp = localtime(&t);
		strftime(timestamp, sizeof(timestamp), "%Y%m%d%H%M%S", tp);
		dg.timestamp = timestamp;
	}
	dg.mc = mc;

	for (; optind < argc; optind++) {
		switch(interval = atoi(argv[optind])) {
		case 300:	/* 5 minutes ==> 1 day of data		*/
			dg.suffix = "day";
			break;
		case 1800:	/* 30 minutes ==> 1 week of data	*/
			dg.suffix = "week";
			break;
		case 7200:	/* 2 hours ==> 1 month of data		*/
			dg.suffix = "month";
			break;
		case 86400:	/* 1 day ==> 1 year of data		*/
			dg.suffix = "year";
			break;
		default:
			mdebug(LOG_CRIT, MDEBUG_LOG, 0, "unknown interval %s",
				argv[optind]);
			exit(EXIT_FAILURE);
			break;
		}
		all_graphs(&dg, interval);
	}

	/* that's all							*/
	exit(EXIT_SUCCESS);
}
